src/open_clip/audio.py

```diff
@@ -0,0 +1,277 @@
+
+        height, width = x.shape[-2:]
+        patch_height, patch_width = self.patch_size
+
+        rounded_height = height // patch_height * patch_height
+        rounded_width = width // patch_width * patch_width
+
+        if (height, width) != (rounded_height, rounded_width):
+            print(f'spectrogram yielded shape of {(height, width)}, but had to be cropped to {(rounded_height, rounded_width)} to be patchified for transformer')
+
+        x = x[..., None, :rounded_height, :rounded_width]
+
+        # pass maybe cropped spectrogram to vit
+
+        return self.vit(x)
+
+# audio class config
+
+@dataclass
+class CLIPAudioCfg(CLIPVisionCfg):
+    spec_n_fft: int = 128
+    spec_power: int = 2
+    spec_win_length: int = 24
+    spec_hop_length: Optional[int] = None
+    spec_pad: int = 0
+    spec_center: bool = True
+    spec_pad_mode: str = 'reflect'
+    aug_stretch_factor: float = 0.8
+    aug_freq_mask: int = 80
+    aug_time_mask: int = 80
+
+# factory method for building audio tower
+
+def _build_audio_tower(
+        embed_dim: int,
+        audio_cfg: CLIPAudioCfg,
+        quick_gelu: bool = False,
+        cast_dtype: Optional[torch.dtype] = None
+):
+    if isinstance(audio_cfg, dict):
+        audio_cfg = CLIPAudioCfg(**audio_cfg)
+
+    # OpenAI models are pretrained w/ QuickGELU but native nn.GELU is both faster and more
+    # memory efficient in recent PyTorch releases (>= 1.10).
+    # NOTE: timm models always use native GELU regardless of quick_gelu flag.
+    act_layer = QuickGELU if quick_gelu else nn.GELU
+
+    audio_heads = audio_cfg.width // audio_cfg.head_width
+    norm_layer = LayerNormFp32 if cast_dtype in (torch.float16, torch.bfloat16) else LayerNorm
+
+    audio = AudioSpectrogramTransformer(
+        image_size=audio_cfg.image_size,
+        patch_size=audio_cfg.patch_size,
+        width=audio_cfg.width,
+        layers=audio_cfg.layers,
+        heads=audio_heads,
+        mlp_ratio=audio_cfg.mlp_ratio,
+        ls_init_value=audio_cfg.ls_init_value,
+        patch_dropout=audio_cfg.patch_dropout,
+        global_average_pool=audio_cfg.global_average_pool,
+        attentional_pool=audio_cfg.attentional_pool,
+        n_queries=audio_cfg.n_queries,
+        attn_pooler_heads=audio_cfg.attn_pooler_heads,
+        output_tokens=audio_cfg.output_tokens,
+        output_dim=embed_dim,
+        act_layer=act_layer,
+        norm_layer=norm_layer,
+        spec_n_fft=audio_cfg.spec_n_fft,
+        spec_power=audio_cfg.spec_power,
+        spec_win_length=audio_cfg.spec_win_length,
+        spec_hop_length=audio_cfg.spec_hop_length,
+        spec_pad=audio_cfg.spec_pad,
+        spec_center=audio_cfg.spec_center,
+        spec_pad_mode=audio_cfg.spec_pad_mode,
+        aug_stretch_factor=audio_cfg.aug_stretch_factor,
+        aug_freq_mask=audio_cfg.aug_freq_mask,
+        aug_time_mask=audio_cfg.aug_time_mask
+    )
+
+    return audio
+
+# audio clip
+
+class AudioCLIP(nn.Module):
```

Comment:
should decide whether to extend `CLIP`

similarly, decide whether to just extend `CoCa`  to `AudioCoCa` and override the visual modality transformer